kernel/src/scan/state.rs

```diff
@@ -1,57 +1,139 @@
+    schema::Schema,
+    DataVisitor, DeltaResult, EngineData, EngineInterface,
+};
 use serde::{Deserialize, Serialize};
 
-pub struct ScanState<'a> {
-    pub(crate) column_types: Vec<ColumnType<'a>>,
-    pub(crate) have_partition_cols: bool,
-    pub(crate) logical_schema: SchemaRef,
-    pub(crate) read_schema: SchemaRef,
+/// State that doesn't change beween scans
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct GlobalScanState {
+    pub table_root: String,
+    pub partition_columns: Vec<String>,
+    pub logical_schema: Schema,
+    pub read_schema: Schema,
 }
 
-impl<'a> ScanState<'a> {
-    pub fn new(
-        column_types: Vec<ColumnType<'a>>,
-        have_partition_cols: bool,
-        logical_schema: SchemaRef,
-        read_schema: SchemaRef,
-    ) -> Self {
-        Self {
-            column_types,
-            have_partition_cols,
-            logical_schema,
-            read_schema,
-        }
-    }
+/// this struct can be used by an engine to materialize a selection vector
+#[derive(Debug)]
+pub struct DvInfo {
+    deletion_vector: Option<DeletionVectorDescriptor>,
+}
 
-    pub fn read_schema(&self) -> SchemaRef {
-        self.read_schema.clone()
+impl DvInfo {
+    pub fn get_selection_vector(
+        &self,
+        engine_interface: &dyn EngineInterface,
+        table_root: &url::Url,
+    ) -> DeltaResult<Option<Vec<bool>>> {
+        let dv_treemap = self
+            .deletion_vector
+            .as_ref()
+            .map(|dv_descriptor| {
+                let fs_client = engine_interface.get_file_system_client();
+                dv_descriptor.read(fs_client, table_root)
+            })
+            .transpose()?;
+        Ok(dv_treemap.map(treemap_to_bools))
     }
+}
 
-    pub fn into_owned(self) -> ScanState<'static> {
-        ScanState {
-            column_types: self
-                .column_types
-                .into_iter()
-                .map(|ct| ct.into_owned())
-                .collect(),
-            have_partition_cols: self.have_partition_cols,
-            logical_schema: self.logical_schema.clone(),
-            read_schema: self.read_schema.clone(),
-        }
-    }
+/// Request that the kernel call a callback on each valid file that needs to be read for the
+/// scan.
+///
+/// The arguments to the callback are:
+/// * `context`: an `&mut context` argument. this can be anything that engine needs to pass through to each call
+/// * `path`: a `&str` which is the path to the file
+/// * `size`: an `i64` which is the size of the file
+/// * `dv_info`: a [`DvInfo`] struct, which allows getting the selection vector for this file
+/// * `partition_values`: a `HashMap<String, String>` which are partition values
+///
+/// ## Context
+/// A note on the `context`. This can be any value the engine wants. This function takes ownership of the passed arg, but then returns it, so the engine can repeatedly call `visit_scan_files` with the same context.
```

Comment:
We need to own it because we put it inside the `ScanFileVisitor` and having that own it just makes all the code much simpler.  I was trying to take a reference and it complicated things, but I can have another crack at it if we really want to try and get that.